### PR TITLE
Finished random channel dropout feature from albumentations

### PR DESCRIPTION
**Pull request by shevdan**
#### Changes

Fixes # (issue)

#### Type of change

- [x] 📚  Documentation Update
- [x] 🧪 Tests Cases
- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+from kornia.augmentation._2d.intensity.base import IntensityAugmentationBase2D
+from kornia.core.check import KORNIA_CHECK
+
+
+class RandomChannelDropout(IntensityAugmentationBase2D):
+    """Randomly Drop Channels in the input Image.
+
+    Args:
+        channel_drop_range (int, int): range from which we choose the number of channels to drop.
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:15]:
```suggestion
        channel_drop_range: range from which we choose the number of channels to drop.
```
types are auto filled in during the docs generation

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        p: probability of applying the transformation.
+        keepdim: whether to keep the output shape the same as input (True) or broadcast it
+                 to the batch form (False).
+
+    Shape:
+        - Input: :math:`(C, H, W)` or :math:`(B, C, H, W)`
+        - Output: :math:`(B, C, H, W)` or :math:`(C, H, W)`
+
+    Examples:
+        >>> import torch
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:27]: import might not needed

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+         [[ True,  True,  True,  True,  True],
+          [ True,  True,  True,  True,  True],
+          [ True,  True,  True,  True,  True],
+          [ True,  True,  True,  True,  True],
+          [ True,  True,  True,  True,  True]]]])
+    """
+
+    def __init__(
+        self,
+        channel_drop_range: Tuple[int, int] = (1, 1),
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:55]:
```suggestion
        channel_drop_range: tuple[int, int] = (1, 1),
```
`from __future__ import annotations` /cc @johnnv1

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
+from random import randint, sample
+from typing import Any, Dict, Optional, Tuple, Union
+
+import numpy as np
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:4]: only pytorch is allowed

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        if len(input.shape) == 3:
+            num_channels = input.shape[0]
+        else:
+            num_channels = input.shape[1]
+        KORNIA_CHECK(self.max_channels < num_channels, "Can not drop all channels in ChannelDropout.")
+        num_drop_channels = randint(self.min_channels, self.max_channels)
+
+        channels_to_drop = sample(range(num_channels), k=num_drop_channels)
+
+        input_cp = clone(input)
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:90]:
```suggestion
        input_cp = input.clone()
```
to avoid import

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
+from random import randint, sample
+from typing import Any, Dict, Optional, Tuple, Union
+
+import numpy as np
+from torch import Tensor, clone
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:5]:
```suggestion
from kornia.core import Tensor
```

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+                (len(input.shape) == 3 and input.shape[0] == 1)
+                or (len(input.shape) == 4 and input.shape[1] == 1)
+                or (len(input.shape) == 2)
+            ),
+            "Only one channel. ChannelDropout is not defined.",
+        )
+        if len(input.shape) == 3:
+            num_channels = input.shape[0]
+        else:
+            num_channels = input.shape[1]
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:84]:
```suggestion
            num_channels = input.shape[-3]
```

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        KORNIA_CHECK(self.max_channels < num_channels, "Can not drop all channels in ChannelDropout.")
+        num_drop_channels = randint(self.min_channels, self.max_channels)
+
+        channels_to_drop = sample(range(num_channels), k=num_drop_channels)
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:88]: you possibly need to implement a batched sampler. See: https://github.com/kornia/kornia/blob/master/kornia/augmentation/_2d/intensity/random_snow.py#L48

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+        input_cp = clone(input)
+        if len(input.shape) == 3:
+            input_cp[channels_to_drop, ...] = self.fill_value
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:92]:
```suggestion
            input_cp[..., channels_to_drop, :, :] = self.fill_value
```
not clear but you could do something like

File: test/augmentation/test_augmentation.py [hunk 10]
```diff
@@ -4134,3 +4135,32 @@ def test_feature_median_blur(self, device, dtype):
...
+
+
+class TestRandomChannelDropout:
+    def test_smoke(self, device, dtype):
+        x_data = torch.rand(1, 3, 5, 5).to(device)
+        aug = RandomChannelDropout(p=0.8)
+        out = aug(x_data)
+        assert out.shape == x_data.shape
+
+    def test_feature_median_blur(self, device, dtype):
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:4147]: this name is wrong

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        )
+        num_channels = input.shape[-3]
+        KORNIA_CHECK(self.max_channels < num_channels, "Can not drop all channels in ChannelDropout.")
+        num_drop_channels = rg.PlainUniformGenerator(((self.min_channels, self.max_channels), "factor_1", None, None))(
+            torch.Size([1])
+        )["factor_1"]
+
+        channels_to_drop = rg.PlainUniformGenerator(((0, num_channels), "factor_1", None, None))(
+            torch.Size([int(torch.round(num_drop_channels).item())])
+        )["factor_1"]
```
Comment (johnnv1) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:85]: the random generator should be initialized just one time -- cc @shijianjian

File: test/augmentation/test_augmentation.py [hunk 12]
```diff
@@ -4187,3 +4188,42 @@ def test_feature_median_blur(self, device, dtype):
         )
 
         utils.assert_close(out, expected)
+
+
+class TestRandomChannelDropout:
```
Comment (johnnv1) [inline test/augmentation/test_augmentation.py:4193]: from kornia.testing import BaseTester
```suggestion
class TestRandomChannelDropout(BaseTester):
```

Check out more about our test guide on https://github.com/kornia/kornia/blob/master/CONTRIBUTING.md#coding-standards

File: test/augmentation/test_augmentation.py [hunk 13]
```diff
@@ -4187,3 +4188,42 @@ def test_feature_median_blur(self, device, dtype):
...
+    def test_smoke(self, device, dtype):
+        x_data = torch.rand(1, 3, 5, 5, dtype=dtype).to(device)
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:4195]:
```suggestion
        x_data = torch.rand(1, 3, 5, 5, dtype=dtype, device=device)
```

File: test/augmentation/test_augmentation.py [hunk 14]
```diff
@@ -4187,3 +4188,42 @@ def test_feature_median_blur(self, device, dtype):
...
+                    [[0.4963, 0.7682], [0.0885, 0.1320]],
+                    [[0.3074, 0.6341], [0.4901, 0.8964]],
+                    [[0.0000, 0.0000], [0.0000, 0.0000]],
+                ]
+            ],
+            device=device,
+            dtype=dtype,
+        )
+
+        utils.assert_close(out, expected)
```
Comment (edgarriba) [inline test/augmentation/test_augmentation.py:4229]:
```suggestion
        self.assert_close(out, expected)
```
After subclassing BaseTester

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        channels_to_drop = list(map(lambda t: t.to(torch.int), channels_to_drop))
+        input_cp = input.clone()
```
Comment (edgarriba) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:87]: Is it really necessary to make a copy ?

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py (hunk 11 above)
Comment (edgarriba) [reply kornia/augmentation/_2d/intensity/random_channel_dropout.py:85]: +1

Comment (edgarriba) [reply kornia/augmentation/_2d/intensity/random_channel_dropout.py:85]: Actually would be great to support integers sampling

**Review by shijianjian — CHANGES_REQUESTED** [review]
Please refer to other files to make a proper use of random generators. They shall not be initialised all the time. 

Also, factor_1 is not a meaningful name.

**Comment by shevdan** [comment]
Fixed all the comments, should be good now. Will appreciate your feedback!

**Review by shevdan — COMMENTED** [review]
Finished refactoring

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+            (input.shape[-3] != 1 and len(input.shape) != 2), "Only one channel. ChannelDropout is not defined."
+        )
+        num_channels = input.shape[-3]
+        KORNIA_CHECK(self.max_channel < num_channels, "Can not drop all channels in ChannelDropout.")
+
+        num_channels_to_drop = torch.randint(low=self.min_channel, high=(self.max_channel + 1), size=(1,))[
+            0
+        ]  # +1 because highest integer to be drawn is not included
+
+        channels_to_drop = torch.randint(low=0, high=num_channels, size=(num_channels_to_drop,))
```
Comment (johnnv1) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:82]: i belive this should be from a random generator or be at least in the `generate_parameter` method like in https://github.com/kornia/kornia/blob/master/kornia/augmentation/_2d/intensity/motion_blur.py#L83

@shijianjian can point out how to do this more correctly
```suggestion
```

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+
+import torch
+
+from kornia.augmentation._2d.intensity.base import IntensityAugmentationBase2D
+from kornia.core import Tensor
+from kornia.core.check import KORNIA_CHECK
+
+
+class RandomChannelDropout(IntensityAugmentationBase2D):
+    """Randomly Drop Channels in the input Image.
```
Comment (shijianjian) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:11]: new line

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+    Args:
+        channel_drop_range: range from which we choose the number of channels to drop.
+        fill_value (int, float): pixel value for the dropped channel.
+        same_on_batch: apply the same transformation across the batch.
+        p: probability of applying the transformation.
+        keepdim: whether to keep the output shape the same as input (True) or broadcast it
+                 to the batch form (False).
+    Shape:
```
Comment (shijianjian) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:19]: new line

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+        - Input: :math:`(C, H, W)` or :math:`(B, C, H, W)`
+        - Output: :math:`(B, C, H, W)` or :math:`(C, H, W)`
+    Examples:
```
Comment (shijianjian) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:22]: new line

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+        )
+        self.min_channel = channel_drop_range[0]
+        self.max_channel = channel_drop_range[1]
+
+    def generate_parameters(self, batch_shape: Tuple[int, ...]) -> Dict[str, Tensor]:
+        params = super().generate_parameters(batch_shape)
+        # +1 to avoid possible error, when low = high
+        params["channel_params"] = torch.randint(low=batch_shape[0], high=batch_shape[1] + 1, size=(batch_shape[2],))
+
+        return params
```
Comment (shijianjian) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:74]: why not using
```suggestion
    self._param_generator =  rg.PlainUniformGenerator((self. channel_drop_range, "channel_params", None, None))
```

Or implement your own like:
https://github.com/kornia/kornia/blob/master/kornia/augmentation/random_generator/_2d/gaussian_blur.py

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+    ) -> Tensor:
+        KORNIA_CHECK(
+            (input.shape[-3] != 1 and len(input.shape) != 2), "Only one channel. ChannelDropout is not defined."
+        )
+        num_channels = input.shape[-3]
+        KORNIA_CHECK(self.max_channel < num_channels, "Can not drop all channels in ChannelDropout.")
+
+        num_channels_to_drop = torch.randint(low=self.min_channel, high=(self.max_channel + 1), size=(1,))[
+            0
+        ]  # +1 because highest integer to be drawn is not included
```
Comment (shijianjian) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:87]: No randomness shall happen during `apply_transform`. It will not be reproducible. Move it to the random generator.

Weird that it passed the reproducibility test ...

**Review by shijianjian — CHANGES_REQUESTED** [review]

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py (hunk 20 above)
Comment (shevdan) [reply kornia/augmentation/_2d/intensity/random_channel_dropout.py:74]: The problem is to make samples of channels to drop,, based on the number of channels to drop

Comment (shijianjian) [reply kornia/augmentation/_2d/intensity/random_channel_dropout.py:82]: We passed the whole batch_shape to random_generator. You have it.

File: kornia/augmentation/_2d/intensity/random_channel_dropout.py [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+            "Invalid channel_drop_range. Should be tuple of length 2.",
+        )
+        KORNIA_CHECK(
+            1 <= channel_drop_range[0] <= channel_drop_range[1],
+            "Invalid channel_drop_range. Max channel should be greater than lower.",
+        )
+        self.min_channel = channel_drop_range[0]
+        self.max_channel = channel_drop_range[1]
+
+        num_channels_to_drop = torch.randint(low=self.min_channel, high=(self.max_channel + 1), size=(1,))[0]
```
Comment (shijianjian) [inline kornia/augmentation/_2d/intensity/random_channel_dropout.py:72]: Please follow https://github.com/kornia/kornia/blob/master/kornia/augmentation/random_generator/_2d/gaussian_blur.py to implement your own generator.

`num_channels_to_drop` you implemented are randomly selected only during the `__init__`. I do not think it is preferred.
